src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,459 @@
+                //        ", is blocked\n",
+                //        W, V);
+            }
+        }
+
+        if (!recurse_deeper) {
+            // L2
+            if (local_found) {
+                // IGRAPH_CHECK(igraph_i_simple_cycles_unblock_recursive(state, V));
+                IGRAPH_CHECK(igraph_i_simple_cycles_unblock(state, V));
+            } else {
+                for (igraph_integer_t i = 0; i < num_neighbors; ++i) {
+                    igraph_integer_t W = VECTOR(*neighbors)[i];
+                    if (!igraph_vector_int_contains(igraph_adjlist_get(&state->B, W),
+                                                    V)) {
+                        IGRAPH_CHECK(igraph_vector_int_push_back(
+                                         igraph_adjlist_get(&state->B, W), V));
+                    }
+                }
+            }
+
+            IGRAPH_ASSERT(!igraph_vector_int_empty(&state->vertex_stack));
+
+            // unstack v
+            V = igraph_vector_int_pop_back(&state->vertex_stack);
+            if (!igraph_vector_int_empty(&state->edge_stack)) {
+                // can be empty for the starting point.
+                // alternatively, V == S
+                E = igraph_vector_int_pop_back(&state->edge_stack);
+            }
+            // printf("Unstacked v %" IGRAPH_PRId ", e %" IGRAPH_PRId "\n", V, E);
+        }
+    }
+
+    *found = local_found;
+
+    IGRAPH_ASSERT(igraph_stack_int_size(&v_stack) == 0);
+    IGRAPH_ASSERT(igraph_stack_int_size(&e_stack) == 0);
+    IGRAPH_ASSERT(igraph_stack_int_size(&neigh_iteration_progress) == 0);
+    igraph_stack_int_destroy(&neigh_iteration_progress);
+    igraph_stack_int_destroy(&v_stack);
+    igraph_stack_int_destroy(&e_stack);
+    IGRAPH_FINALLY_CLEAN(3);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_simple_cycle_search_state_init
+ * \brief Initialize the cycle search state
+ *
+ * \experimental
+ *
+ * \param state The state structure to initialize
+ * \param graph The graph object
+ * \return Error code.
+ *
+ * Time complexity: O(|V|*|E|*log(|V|*|E|))
+ *
+ * \ref igraph_simple_cycle_search_state_destroy
+ */
+igraph_error_t
+igraph_simple_cycle_search_state_init(igraph_simple_cycle_search_state_t *state,
+                                      const igraph_t *graph) {
+    state->N = igraph_vcount(graph);
+    state->NE = igraph_ecount(graph);
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&state->vertex_stack, 0);
+    IGRAPH_CHECK(igraph_vector_int_reserve(&state->vertex_stack, 8));
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&state->edge_stack, 0);
+    IGRAPH_CHECK(igraph_vector_int_reserve(&state->edge_stack, 8));
+    IGRAPH_CHECK(igraph_vector_bool_init(&state->v_blocked, state->N));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &state->v_blocked);
+    IGRAPH_CHECK(igraph_inclist_init(
+                     graph, &state->IK, IGRAPH_OUT,
+                     IGRAPH_LOOPS_ONCE)); // TODO: understand what we actually want to include
```

Comment:
```suggestion
    IGRAPH_CHECK(igraph_inclist_init(
                     graph, &state->IK, IGRAPH_OUT,
                     IGRAPH_LOOPS_ONCE)); // each self-loop counts as a single cycle in directed graphs
```

I think this is correct, and the TODO note can be removed.